packages/cozy-realtime/src/RetryManager.js

```diff
@@ -0,0 +1,142 @@
+import MicroEE from 'microee'
+
+import {
+  maxWaitBetweenRetries,
+  baseWaitAfterFirstFailure,
+  timeBeforeSuccessfull,
+  raiseErrorAfterAttempts
+} from './config'
+
+class RetryManager {
+  constructor() {
+    this.reset()
+    this.onSuccess = this.onSuccess.bind(this)
+    this.onFailure = this.onFailure.bind(this)
+    this.onSuccessAcknowledged = this.onSuccessAcknowledged.bind(this)
+  }
+
+  /**
+   * On a success
+   *
+   * If `timeBeforeSuccessfull` if not null, it will wait this amount of
+   * time without errors before acknowleding a success
+   */
+  onSuccess() {
+    if (timeBeforeSuccessfull > 0) {
+      this.startSuccessWatcher(
+        this.onSuccessAcknowledged,
+        timeBeforeSuccessfull
+      )
+    } else {
+      this.onSuccessAcknowledged()
+    }
+  }
+
+  /**
+   * Acknowledge a success
+   *
+   * @private
+   */
+  onSuccessAcknowledged() {
+    this.emit('success')
+    this.reset()
+  }
+
+  /**
+   * Clears the success watcher
+   *
+   * @private
+   */
+  clearSuccessWatcher() {
+    if (this.successWatcher) {
+      global.clearTimeout(this.successWatcher)
+      this.successWatcher = null
+    }
+  }
+
+  /**
+   * Starts the success watcher - acknowledges the success if not interrupted after some time
+   *
+   * @private
+   * @param {function} ack - will be called after the wait time
+   * @param {number} time - waiting time
+   */
+  startSuccessWatcher(ack, time) {
+    if (!this.successWatcher) {
+      const callback = () => {
+        this.successWatcher = null
+        ack()
+      }
+      this.successWatcher = global.setTimeout(callback, time)
+    }
+  }
+
+  /**
+   * On failure
+   *
+   * If we where waiting for a success timeout, it will cancel the wait.
+   *
+   * An application error (failed authentication, failed subscription) will
+   * not trigger a failure (trying to reconnect with the same parameters
+   * has no meaning anyways).
+   *
+   * @param {object} error
+   */
+  onFailure(error) {
+    this.clearSuccessWatcher()
+    this.increaseFailureCounter()
+    this.emit('failure', error)
+    if (this.hasReachedErrorLevel()) this.emit('error', error)
+  }
+
+  /**
+   * Returns if we reached a level where we need to raise an error
+   *
+   * @returns {boolean}
+   */
+  hasReachedErrorLevel() {
+    return this.retries > 0 && this.retries % raiseErrorAfterAttempts == 0
+  }
```

Comment:
Architecture: I wonder her why `raiseErrorAfterAttempts` is not a parameter of the class instead of a global config variable.

Naming: hasReachedErrorLevel seems to imply that the "error level" is a threshold whereas modular arithmetic is applied, meaning that the level can be reached multiple times. Maybe `shouldEmitError` could be used instead ?

Reply:
:+1: done